Feature/channel pool validation

itayx opened the pull request:
Currently the DCS's set_channel_scan_pool uses a static list of supported channels for verifications to see if the channels provided in the channel pool are supported.

The DB has a list known as hostap_supported_channels, this is filled with channels provided by the basic radio capabilities

The DCS's set_channel_scan_pool need to validate it's channel pool against this supported channels list

kantera800 commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2299,16 +2302,14 @@ bool db::set_channel_scan_pool(const sMacAddr &mac, const std::unordered_set<uin
+                             [&channel](const beerocks::message::sWifiChannel &supported_channel) {
+                                 return supported_channel.channel == channel;
+                             });
+        if (invalid_channel == supported_channels.end()) {
```
please use separate method instead  `is_channel_scan_pool_supported()` 
also, no need to scan all pool , just  return false for the first channel that is not supported

itayx replied:
added `is_channel_scan_pool_supported()`

itayx committed:
db: use supported channels for channel pool verification

In the set_channel_scan_pool function to validate the channel pool, we
run a check to see if the channel is supported. Currently the supported
channel list is a static list.

Add is_channel_scan_pool_supported function to the DB.
Use the is_supported function to validate channel pool.

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

CoralMalachi commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2289,6 +2290,24 @@ int db::get_channel_scan_dwell_time_msec(const sMacAddr &mac, bool single_scan)
+                                        const std::unordered_set<uint8_t> &channel_pool)
+{
+    auto supported_channels = get_hostap_supported_channels(network_utils::mac_to_string(mac));
+    for (auto channel : channel_pool) {
```
```suggestion
    for (const auto &channel : channel_pool) {
```

As far as I understand, you are only reading the `channel` value and you have no attempt to modify it, so it preferred to make it a read-only value.

itayx replied:
Changed

rmelotte commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2289,6 +2290,24 @@ int db::get_channel_scan_dwell_time_msec(const sMacAddr &mac, bool single_scan)
+                             return supported_channel.channel == channel;
+                         });
+        if (found_channel == supported_channels.end()) {
+            LOG(ERROR) << "channel #" << int(channel) << " is not supported";
```
To me it's a bit weird that a function called `is_channel_scan_pool_supported` logs something (especially an error) if a channel is not supported. Maybe it would be best to leave that to the caller (or to change the level to INFO or DEBUG).

Aren't there any case where you would want to call `is_channel_scan_pool_supported` without logging an error?

(nitpick)

itayx replied:
I completely understand
And I think I'll change it to `INFO`
But I still want the log, that way there is a log of which sent channel isn't supported
Since the function doesn't return the unsupported channel, the only way to know which is via the log

rmelotte commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -512,6 +512,17 @@ class db {
     bool set_channel_scan_pool(const sMacAddr &mac, const std::unordered_set<uint8_t> &channel_pool,
                                bool single_scan);
 
+    /**
+     * @brief Validate the channel scan pool
+     * 
+     * @param mac:          MAC address of radio
+     * @param channel_pool: Channel pool of channel scan
+     * @return true if pool is valid
+     * @return false if pool is invalid
```
To me the function name is pretty explicit about what is being done already, but the documentation is so vague that it doesn't bring anything useful. 
What kind of validation does it do?

For example:

```suggestion
    /**
     * @brief Validate the channel scan pool.
     * 
     * A pool is considered valid if all the channels it contains are supported by the radio.
     * 
     * @param mac:          MAC address of the radio
     * @param channel_pool: Channel pool to validate
     * @return true if channel_pool is valid, false otherwise.
```

(nitpick)

itayx replied:
Updated the doxygen documentation

rmelotte reviewed (approved):
Only nitpicks :+1:

kantera800 commented:
approved , but don't merge it yet.
**Wait for the NO_IR issue to be solved.**

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -1248,8 +1248,9 @@ std::string db::get_hostap_supported_channels_string(const std::string &radio_ma
-               << " [dbm] | bss_overlap = " << int(val.bss_overlap) << std::endl;
+               << " | tx_pow = " << int(val.tx_pow) << " | noise = " << int(val.noise) << " [dbm]"
+               << " | bss_overlap = " << int(val.bss_overlap)
+               << " | bw = " << int(val.channel_bandwidth) << std::endl;
```
This should be a separate commit (nitpick)

itayx replied:
This was moved to #1176 
I will remove it from here once that PR is merged
